src/tonations_generation.py

```diff
@@ -0,0 +1,203 @@
+        chord_type = sound_to_chord_type(
+            sound, tonic, subdominant, dominant, tend_to_tonic)
+        tend_to_tonic = chord_type != ChordType.SUBDOMINANT
+        chord_types.append(chord_type)
+    return chord_types
+
+
+def sound_to_chord_type(sound, tonic, subdominant, dominant, tend_to_tonic):
+    if sound.note in dominant:
+        return ChordType.DOMINANT
+    if sound.note in tonic:
+        if tend_to_tonic or sound.note not in subdominant:
+            return ChordType.TONIC
+        return ChordType.SUBDOMINANT
+    if sound.note in subdominant:
+        return ChordType.SUBDOMINANT
+    return ChordType.OTHER
+
+
+def remove_solitary_chords(chord_types):
+    if len(chord_types) < 2:
+        return chord_types
+    result = []
+    if chord_types[0] == chord_types[1]:
+        result.append(chord_types[0])
+    for i in range(1, len(chord_types)-1):
+        chord = chord_types[i]
+        if chord_types[i-1] == chord or chord == chord_types[i+1]:
+            result.append(chord)
+    if chord_types[-1] == chord_types[-2]:
+        result.append(chord_types[-1])
+    return result
+
+
+def join_same_consecutive_chords(chord_types):
+    if len(chord_types) < 2:
+        return chord_types
+    result = []
+    last_chord = chord_types[0]
+    result.append(last_chord)
+    for chord in chord_types:
+        if chord != last_chord:
+            last_chord = chord
+            result.append(chord)
+    return result
+
+
+def get_points_from_scale_fit(sounds, scale):
+    points = 0
+    for sound in sounds:
+        if sound.note in scale:
+            points += 10
+    return points
+
+
+class ChordSequencePattern:
+    def __init__(self, sequence, points):
+        self.sequence = sequence
+        self.points = points
+
+
+def get_points_from_chord_patterns_fit(chords):
+    patterns = []
+    patterns.append(ChordSequencePattern(
+        [ChordType.TONIC, ChordType.SUBDOMINANT,
+            ChordType.DOMINANT, ChordType.TONIC],
+        POINTS_CADENCE
+    ))
+    patterns.append(ChordSequencePattern(
+        [ChordType.SUBDOMINANT, ChordType.DOMINANT, ChordType.TONIC],
+        POINTS_GREAT_AUTHENTIC_CADENCE
+    ))
+    patterns.append(ChordSequencePattern(
+        [ChordType.TONIC, ChordType.DOMINANT, ChordType.TONIC],
+        POINTS_AUTHENTIC_CADENCE
+    ))
+    patterns.append(ChordSequencePattern(
+        [ChordType.TONIC, ChordType.SUBDOMINANT, ChordType.TONIC],
+        POINTS_PLAGAL_CADENCE
+    ))
+    patterns.append(ChordSequencePattern(
+        [ChordType.SUBDOMINANT, ChordType.DOMINANT],
+        POINTS_HALF_CADENCE
+    ))
+    patterns.append(ChordSequencePattern(
+        [ChordType.TONIC, ChordType.DOMINANT],
+        POINTS_HALF_CADENCE
+    ))
+    # we don't use deceptive cadence, as it was not in the book
+    points = 0
+    for pattern in patterns:
+        no_occurences = get_no_pattern_occurences(chords, pattern.sequence)
+        points += no_occurences * pattern.points
```

Comment:
Point calculation does not work, because eg HALF_CADENCE is already included in PLAGAL_CADENCE so that everytime there is pattern PLAGAL_CADENCE it will also reward points for HALF_CADENCE.
Fortunately it doesn't matter, because we can mitigate it using constants

Reply:
You're right. I think adjusting constants seem to be the best way out